### ISSUE TITLE
Add and refine unit tests for autofill functions

### ISSUE THREAD

**Issue by jinfu93** [JMSLab/Template#55]
Porting over from the [sketch](https://github.com/JMSLab/Template/pull/51#discussion_r827376226) in #51 below are the additional tests we would like to implement in `test_autofill.py`:

- [x] Test that the function creates a new file.
- [x] Test that the file created by the function contains the correct command names and values.
- [x] What is the expected behavior of the function when a variable name is not found?
  - Add error handling.
  - Construct a test to verify error handling works as expected.
- [x] Make sure the behavior is as expected when using a string for format and a list with different formats.
  - Should we allow for a default format, so that the function knows what to do when `autofill_formats` is not specified?
- [x] Streamline test suite

@santiagohermo do feel free to add to this list or expand on the details of each test, thanks a lot!

**Comment by santiagohermo** [comment]
> @santiagohermo do feel free to add to this list or expand on the details of each test, thanks a lot!

Thanks @jinfu93! I made a couple of edits to the [opening comment](https://github.com/JMSLab/Template/issues/55#issue-1181188969).

**Comment by jinfu93** [comment]
@santiagohermo I've added drafts of 2 tests:
https://github.com/JMSLab/Template/blob/442eb18e64e6348e8768987c238b150b9140c874/source/lib/JMSLab/tests/test_autofill.py#L26-L44

Let me know if these look different from what you had in mind!

Regarding:
>Make sure the behavior is as expected when using a string for format and a list with different formats.

I'm thinking of adding error handling to accommodate when `autofill_formats` doesn't line up with `autofill_lists` i.e. if `autofill_lists` is a nested list but `autofill_formats` is a string. Subsequently, the test can ensure the exception is thrown as expected. Let me know if you were thinking of something more straightforward than that.

I've also added a default for `autofill_formats` in [1a52e3e](https://github.com/JMSLab/Template/commit/1a52e3e3b47d2031c2a7ad728a3d6144ff9adb4e) as you suggested.

Thanks!

**Comment by jinfu93** [comment]
@santiagohermo implemented the following:
>I'm thinking of adding error handling to accommodate when autofill_formats doesn't line up with autofill_lists i.e. if autofill_lists is a nested list but autofill_formats is a string. Subsequently, the test can ensure the exception is thrown as expected. Let me know if you were thinking of something more straightforward than that.

in [a5e665e](https://github.com/JMSLab/Template/commit/a5e665e5a3917cf98c7bcbfa4a2a665115a72067) and [55c431a](https://github.com/JMSLab/Template/commit/55c431ac9d210cf0ab6231698b3ee5bc0c1eb29c) fyi. Thanks again!

**Comment by santiagohermo** [comment]
Thanks @jinfu93! On a first look the scripts look great. One small comment is that I'd try to not use [`sys.path.append`](https://github.com/JMSLab/Template/blob/55c431ac9d210cf0ab6231698b3ee5bc0c1eb29c/source/lib/JMSLab/tests/test_autofill.py#L3) to import files in upstream folders. I'd use the approach [here](https://github.com/JMSLab/Template/blob/55c431ac9d210cf0ab6231698b3ee5bc0c1eb29c/source/lib/JMSLab/tests/test_build_lyx.py#L10-L15) instead.

Feel free to move to PR and I'll make a more detailed review there.

**Comment by jinfu93** [comment]
Thanks @santiagohermo for the great tip! I attempted implementing the approach you suggested locally but was met with:

>from ..autofill import GenerateAutofillMacros
ImportError: attempted relative import with no known parent package

I also tried running locally the script `source/lib/JMSLab/tests/test_build_lyx.py` that you referenced but got the same error. I'm wondering if you might have some idea about whether that's a potential setup/configuration issue on my end? Thanks!

**Comment by santiagohermo** [comment]
Thanks @jinfu93! What was the working directory when you ran the script?

**Comment by jinfu93** [comment]
@santiagohermo I tried running it from both the root of Template and from the location of the script.

**Comment by santiagohermo** [comment]
Can you try running it from `source/lib/JMSLab`? When I run `pytest` from that folder things seem to work fine.

**Comment by jinfu93** [comment]
Thanks @santiagohermo! I could run `pytest` from `sources/lib/JMSLab` but running the script itself without `pytest` would still throw out the same error. Nonetheless, I could do some local testing with `pytest` which is more than sufficient. Thanks again!

**Comment by jinfu93** [comment]
This thread continues in its PR #57.

**Comment by jinfu93** [comment]
Summary:

In this issue, we completed the suite of unit tests for the new autofill function in Python along with its accompanying wrapper function.

Changes were merged into `main` in [35b4fef](https://github.com/JMSLab/Template/commit/35b4fefa342708e080cf1a6736183da0c895580f).

Final state of issue branch can be viewed [here](https://github.com/JMSLab/Template/tree/fff134e5d649825bf81801c2a80e2acb8f8d2c0c).
